pkg/storage/stores/series/series_index_store.go

```diff
@@ -244,31 +243,57 @@ func (c *indexReaderWriter) chunksToSeries(ctx context.Context, in []logproto.Ch
 		}
 	}
 
-	for _, group := range groups {
-		sort.Sort(group)
-		chunks, err := c.fetcher.FetchChunks(ctx, group.chunks, group.keys)
-		if err != nil {
-			return nil, err
-		}
+	type f func() ([]labels.Labels, error)
+	var jobs []f
 
-	outer:
-		for _, chk := range chunks {
-			for _, matcher := range matchers {
-				if matcher.Name == astmapper.ShardLabel || matcher.Name == labels.MetricName {
-					continue
+	for _, g := range groups {
+    group := g
+		jobs = append(jobs, f(func() ([]labels.Labels, error) {
+			sort.Sort(group)
+			chunks, err := c.fetcher.FetchChunks(ctx, group.chunks, group.keys)
+			if err != nil {
+				return nil, err
+			}
+
+		results := make([]labels.Labels, 0, len(chunks))
+		outer:
+			for _, chk := range chunks {
+				for _, matcher := range matchers {
+					if matcher.Name == astmapper.ShardLabel || matcher.Name == labels.MetricName {
+						continue
+					}
+					if !matcher.Matches(chk.Metric.Get(matcher.Name)) {
+						continue outer
+					}
 				}
-				if !matcher.Matches(chk.Metric.Get(matcher.Name)) {
+
+				if chunkFilterer != nil && chunkFilterer.ShouldFilter(chk.Metric) {
 					continue outer
 				}
-			}
 
-			if chunkFilterer != nil && chunkFilterer.ShouldFilter(chk.Metric) {
-				continue outer
+				results = append(results, labels.NewBuilder(chk.Metric).Del(labels.MetricName).Labels())
 			}
 
-			results = append(results, labels.NewBuilder(chk.Metric).Del(labels.MetricName).Labels())
-		}
+			return results, nil
+		}))
+	}
+
+	results := make([]labels.Labels, 0, len(chunksBySeries))
+	if err := concurrency.ForEachJob(
+		ctx,
+		len(jobs),
```

Comment:
wdyt of putting a upper bound to the concurrency number? like: min(len(jobs), 20)